clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -57,238 +56,39 @@
  *     the fetch offset after the index lookup
  * - {@link Errors#UNKNOWN_SERVER_ERROR} For any unexpected errors
  */
-public class FetchResponse<T extends BaseRecords> extends AbstractResponse {
-
-    public static final long INVALID_HIGHWATERMARK = -1L;
+public class FetchResponse extends AbstractResponse {
+    public static final long INVALID_HIGH_WATERMARK = -1L;
     public static final long INVALID_LAST_STABLE_OFFSET = -1L;
     public static final long INVALID_LOG_START_OFFSET = -1L;
     public static final int INVALID_PREFERRED_REPLICA_ID = -1;
 
     private final FetchResponseData data;
-    private final LinkedHashMap<TopicPartition, PartitionData<T>> responseDataMap;
+    // we build responseData when needed.
+    private LinkedHashMap<TopicPartition, FetchResponseData.PartitionData> responseData = null;
```

Comment:
This is not thread-safe and requests are typically thread-safe. What's the thinking here?

Reply:
Looks like this method is used only `Fetcher.sendFetches` outside of tests, benchmaks, etc.. Also once in `errorCounts`, but we can change that code. Maybe we can remove this altogether. Thoughts?

Reply:
I think I would move the method to a test utility so that tests can use that instead.

Reply:
It is used by `FetchSessionHandler` also. I prefer to rewrite that code by another PR to avoid big patch :(

It seems to me using synchronization block can resolve the thread issue. Also, I will file a jira as follow-up. WDYT?

Reply:
Yeah, you can use a volatile field and synchronize on the assignment if still null. And then file a separate Jira to remove it from the class in a separate PR.

Reply:
jira: https://issues.apache.org/jira/browse/KAFKA-12385

> you can use a volatile field and synchronize on the assignment if still null.

updated